src/metabase/api/database.clj

```diff
@@ -414,6 +416,41 @@
      :order-by [[:%lower.name :asc]]
      :limit    limit}))
 
+(defn- autocomplete-cards
+  "Returns cards that match the search string in the given database, ordered by id.
+  `search-card-slug` should be in a format like '123-foo-bar' or '123' or 'foo-bar', where 123 is the card ID
+   and foo-bar is a prefix of the card name converted into a slug.
+
+   If the search string contains a number like '123' we match that as a prefix against the card IDs.
+   If the search string contains a number at the start AND text like '123-foo' we match do an exact match on card ID, and a substring match on the card name.
+   If the search string does not start with a number, and is text like 'foo' we match that as a substring on the card name."
+  [database-id search-card-slug]
+  (let [search-id   (re-find #"\d*" search-card-slug)
+        search-name (-> (re-matches #"\d*-?(.*)" search-card-slug)
+                        second
+                        (str/replace #"-" " ")
+                        str/lower-case)]
+    (db/select [Card :id :dataset :database_id :name :collection_id]
+               {:where    [:and
+                           [:= :database_id database-id]
+                           [:= :archived false]
+                           (cond
+                             ;; e.g. search-string = "123"
+                             (and (not-empty search-id) (empty? search-name))
+                             [:like (hx/cast (if (= (mdb.connection/db-type) :mysql) :char :text) :id) (str search-id "%")]
+
+                             ;; e.g. search-string = "123-foo"
+                             (and (not-empty search-id) (not-empty search-name))
+                             [:and
+                              [:= :id (Integer/parseInt search-id)]
+                              [:like :%lower.name (str "%" search-name "%")]]
+
+                             ;; e.g. search-string = "foo"
+                             (and (empty? search-id) (not-empty search-name))
+                             [:like :%lower.name (str "%" search-name "%")])]
```

Comment:
if you lower the name should you also lower case the `search-name`? Seems like a confusing way to search for "Products" and it not match on "Products" because it is checking against "products".

Reply:
it's already lowered in the let form.